File: srcs/record_cam.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <string>

using namespace cv;
using namespace std;

int main()
{
    FileStorage fs("calibration.yml", FileStorage::READ);
    if (!fs.isOpened())
    {
        cerr << "Failed to open calibration.yml" << endl;
        return -1;
    }
    Mat cameraMatrix, distCoeffs;
    fs["CameraMatrix"] >> cameraMatrix;
    fs["DistCoeffs"] >> distCoeffs;
    fs.release();

    string pipeline =
        "nvarguscamerasrc sensor-id=0 ! "
        "video/x-raw(memory:NVMM), width=640, height=480, format=NV12, "
        "framerate=15/1 ! "
        "nvvidconv ! video/x-raw, format=BGRx ! "
        "videoconvert ! video/x-raw, format=BGR ! "
        "appsink";

    VideoCapture cap(pipeline, CAP_GSTREAMER);
    if (!cap.isOpened())
    {
        cerr << "Error: Could not open the CSI camera." << endl;
        return -1;
    }

    Mat frame, undistorted;
    cap >> frame;
    if (frame.empty())
    {
        cerr << "Error: Could not capture initial frame." << endl;
        return -1;
    }

    int width = frame.cols;
    int height = frame.rows;

    // Setup video writer
    VideoWriter writer("video.avi", VideoWriter::fourcc('X','V','I','D'), 15, Size(width, height));
    if (!writer.isOpened())
    {
        cerr << "Error: Could not open video writer." << endl;
        return -1;
    }

    namedWindow("Undistorted Feed", WINDOW_AUTOSIZE);

    cout << "Recording... Press 'q' to stop and save video." << endl;

    while (true)
    {
        if (!cap.read(frame))
        {
            cerr << "Error: Unable to capture frame." << endl;
            break;
        }

        undistort(frame, undistorted, cameraMatrix, distCoeffs);
        writer.write(undistorted);
        imshow("Undistorted Feed", undistorted);

        char key = (char)waitKey(1);
        if (key == 'q' || key == 27) // 'q' or ESC
        {
            break;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(2));
    }

    cap.release();
    writer.release();
    destroyAllWindows();

    cout << "Video saved as 'video.avi'" << endl;
    return 0;
}

File: record_cam.py
import cv2
import numpy as np

# Load calibration parameters
fs = cv2.FileStorage("calibration.yml", cv2.FILE_STORAGE_READ)
camera_matrix = fs.getNode("CameraMatrix").mat()
dist_coeffs = fs.getNode("DistCoeffs").mat()
fs.release()

# Open the camera
cap = cv2.VideoCapture(0)
cap.set(cv2.CAP_PROP_FRAME_WIDTH, 1280)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 720)

if not cap.isOpened():
    print("Error: Could not open camera.")
    exit()

# Grab a sample frame to get dimensions
ret, frame = cap.read()
if not ret:
    print("Error: Could not read frame.")
    cap.release()
    exit()

h, w = frame.shape[:2]

# Compute undistortion map
new_camera_matrix, roi = cv2.getOptimalNewCameraMatrix(
    camera_matrix, dist_coeffs, (w, h), 1, (w, h)
)
map1, map2 = cv2.initUndistortRectifyMap(
    camera_matrix, dist_coeffs, None, new_camera_matrix, (w, h), cv2.CV_16SC2
)

# Setup video writer
fourcc = cv2.VideoWriter_fourcc(*'XVID')
out = cv2.VideoWriter('calibrated_output.avi', fourcc, 20.0, (w, h))

print("Recording calibrated video... Press Ctrl+C to stop.")

try:
    while True:
        ret, frame = cap.read()
        if not ret:
            print("Failed to grab frame.")
            break

        # Apply undistortion
        undistorted = cv2.remap(frame, map1, map2, interpolation=cv2.INTER_LINEAR)

        # Write to file
        out.write(undistorted)

        # Optional: show preview
        # cv2.imshow('Calibrated Video', undistorted)
        # if cv2.waitKey(1) & 0xFF == ord('q'):
        #     break

except KeyboardInterrupt:
    print("\nStopped by user.")

# Cleanup
cap.release()
out.release()
cv2.destroyAllWindows()
print("Video saved as 'video1.avi'")
